src/estimation.jl

```diff
@@ -38,4 +40,155 @@ function esprit(x::AbstractArray, M::Integer, p::Integer, Fs::Real=1.0)
+    f = fftshift(fftfreq(N, Fs))[k]  # frequency at index k
+    if (k+1 <= N) && (k-1 >= 1)
+        δ = -real((X[k+1] - X[k-1]) / (2X[k] - X[k-1] - X[k+1]) )
+    else
+        δ = 0.0
+    end
+
+    f + δ*Fs/N
+end
+
+"""
+    quinn(x::Vector, f0::Real = 0.0, Fs::Real = 1.0 ; tol = 1e-6, maxiters = 20)
+
+    quinn(x::Vector, Fs::Real = 1.0 ; kwargs...)
+
+    quinn(x::Vector ; kwargs...)
+
+Algorithms by Quinn and Quinn & Fernandes for frequency estimation. Given a
+signal `x` and an initial guess `f0`, estimate and return the frequency of the
+largest sinusoid in `x`. `Fs` is the sampling frequency. All frequencies are
+expressed in Hz.
+
+If the initial guess `f0` is not provided or if it is equal to zero, then a guess
+is calculated by `quinn` using Jacobsen's estimator. The sampling frequency `Fs`
+defaults to `1.0`.
+
+The following keyword arguments control the algorithm's behavior:
+
+- `tol`: () the algorithm stops when the absolut value of the
+  difference between two consecutive estimates is less than `tol`. Defaults to
+  `1e-6`.
+- `maxiters`: the maximum number of iterations to run. Defaults to `20`.
+
+Returns a tuple `(estimate, reachedmaxiters)`, where `estimate` is the
+estimated frequency, and `reachedmaxiters` is `true` if the algorithm finished
+after running for `maxiters` iterations (this may indicate that the algorithm
+did not converge).
+
+If the signal `x` is real, then the algorithm used is [^Quinn1991]. If the signal is
+complex, the algorithm is [^Quinn2009].
+
+[^Quinn1991]: B Quinn and J Fernandes, "A fast efficient technique for the
+estimation of frequency", Biometrika, Vol. 78 (1991).
+
+[^Quinn2009]: B Quinn, "Recent advances in rapid frequency estimation", Digital
+Signal Processing, Vol. 19 (2009), Elsevier.
+
+"""
+quinn(x ; kwargs...) = quinn(x, jacobsen(x, Fs), 1.0 ; kwargs...)
+
+quinn(x, Fs ; kwargs...) = quinn(x, jacobsen(x, Fs), Fs ; kwargs...)
+
+function quinn(x::Vector{<:Real}, f0::Real, Fs::Real ; tol = 1e-6, maxiters = 20)
+    fₙ = Fs/2
+    T = length(x)
+
+    # Run a quick estimate of largest sinusoid in x
+    ω̂ = π*f0/fₙ
+
+    # remove DC
+    x .= x .- mean(x)
+
+    # Initialize algorithm
+    α = 2cos(ω̂)
+
+    # iteration
+    ξ = zeros(eltype(x), T)
+    β = zero(eltype(x))
+    iter = 0
+    @inbounds while iter < maxiters
+        iter += 1
+        ξ[1] = x[1]
+        ξ[2] = x[2] + α*ξ[1]
+        for t in 3:T
+            ξ[t] = x[t] + α*ξ[t-1] - ξ[t-2]
+        end
+        β = ξ[2]/ξ[1]
+        for t = 3:T
+            β += (ξ[t]+ξ[t-2])*ξ[t-1]
+        end
+        β = β/(ξ[1:end-1]'*ξ[1:end-1])
+        abs(α - β) < tol && break
+        α = 2β-α
+    end
+
+    fₙ*acos(0.5*β)/π, iter == maxiters
+end
+
+function quinn(x::Vector{<:Complex}, f0::Real, Fs::Real ; tol = 1e-6, maxiters = 20)
+    fₙ = Fs/2
+    T = length(x)
+
+    ω̂ = π*f0/fₙ
+
+    # Remove any DC term in x
+    x .= x .- complex(mean(real.(x)), mean(imag.(x)))
```

Comment:
Why not
```suggestion
    x .= x .- mean(x)
```
?

Reply:
Fixed -- I either missed this method of `mean`, or it didn't exist when I first wrote this code.